File: apps/hypercortex/src/assetFileTypes.ts
import assetFileTypes from '../backend-go/shared/asset_file_types.json'

type AssetFileKind = 'image' | 'audio' | 'video' | 'document'

type AssetFileType = {
  ext: string
  mime: string
  kind: AssetFileKind
  mimeAliases?: readonly string[]
}

const ASSET_FILE_TYPES = loadAssetFileTypes(assetFileTypes as readonly AssetFileType[])

const ASSET_FILE_TYPE_BY_EXT = new Map(ASSET_FILE_TYPES.map(item => [item.ext, item]))
const EXT_BY_MIME = new Map<string, string>()
const KIND_BY_MIME = new Map<string, AssetFileKind>()

for (const item of ASSET_FILE_TYPES) {
  setPreferredMimeExt(item.mime, item.ext)
  setPreferredMimeKind(item.mime, item.kind)
  for (const alias of item.mimeAliases || []) setPreferredMimeExt(alias, item.ext)
  for (const alias of item.mimeAliases || []) setPreferredMimeKind(alias, item.kind)
}

export function extFromMime(mime: string): string {
  return EXT_BY_MIME.get(normalizeMime(mime)) || ''
}

export function mimeFromExt(ext: string): string {
  return ASSET_FILE_TYPE_BY_EXT.get(normalizeExt(ext))?.mime || ''
}

export function kindFromMime(mime: string): string {
  const m = normalizeMime(mime)
  const catalogKind = KIND_BY_MIME.get(m)
  if (catalogKind) return catalogKind
  if (m.startsWith('image/')) return 'image'
  if (m.startsWith('audio/')) return 'audio'
  if (m.startsWith('video/')) return 'video'
  return 'document'
}

export const ACCEPTED_FILE_EXTENSIONS = ASSET_FILE_TYPES.map(item => item.ext)

function normalizeExt(ext: string): string {
  return String(ext || '').toLowerCase().replace(/^\./, '').trim()
}

function normalizeMime(mime: string): string {
  return String(mime || '').toLowerCase().split(';')[0].trim()
}

function setPreferredMimeExt(mime: string, ext: string): void {
  const normalizedMime = normalizeMime(mime)
  if (!normalizedMime || EXT_BY_MIME.has(normalizedMime)) return
  EXT_BY_MIME.set(normalizedMime, ext)
}

function setPreferredMimeKind(mime: string, kind: AssetFileKind): void {
  const normalizedMime = normalizeMime(mime)
  if (!normalizedMime || KIND_BY_MIME.has(normalizedMime)) return
  KIND_BY_MIME.set(normalizedMime, kind)
}

function loadAssetFileTypes(items: readonly AssetFileType[]): readonly AssetFileType[] {
  if (!items.length) throw new Error('asset file types cannot be empty')
  const seenExts = new Set<string>()
  for (const item of items) {
    const ext = normalizeExt(item.ext)
    const mime = normalizeMime(item.mime)
    if (!ext || !mime || !isAssetFileKind(item.kind)) throw new Error(`invalid asset file type: ${JSON.stringify(item)}`)
    if (seenExts.has(ext)) throw new Error(`duplicate asset file extension: ${ext}`)
    seenExts.add(ext)
  }
  return items
}

function isAssetFileKind(kind: unknown): kind is AssetFileKind {
  return kind === 'image' || kind === 'audio' || kind === 'video' || kind === 'document'
}

File: apps/hypercortex/backend-go/asset_file_types.go
package main

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"strings"
)

//go:embed shared/asset_file_types.json
var assetFileTypesJSON []byte

type assetFileType struct {
	Ext         string   `json:"ext"`
	Mime        string   `json:"mime"`
	Kind        string   `json:"kind"`
	MimeAliases []string `json:"mimeAliases"`
}

var assetFileTypes = mustLoadAssetFileTypes()

var assetFileTypeByExt = buildAssetFileTypeByExt(assetFileTypes)
var assetKindByMime = buildAssetKindByMime(assetFileTypes)

func isAllowedAssetExt(ext string) bool {
	_, ok := assetFileTypeByExt[normalizeAssetFileExt(ext)]
	return ok
}

func assetFileMimeFromExt(ext string) string {
	if item, ok := assetFileTypeByExt[normalizeAssetFileExt(ext)]; ok {
		return item.Mime
	}
	return ""
}

func assetFileKindFromMime(mimeType string) string {
	return assetKindByMime[normalizeAssetMime(mimeType)]
}

func buildAssetFileTypeByExt(items []assetFileType) map[string]assetFileType {
	out := map[string]assetFileType{}
	for _, item := range items {
		out[normalizeAssetFileExt(item.Ext)] = item
	}
	return out
}

func buildAssetKindByMime(items []assetFileType) map[string]string {
	out := map[string]string{}
	for _, item := range items {
		kind := strings.TrimSpace(item.Kind)
		setPreferredAssetMimeKind(out, item.Mime, kind)
		for _, alias := range item.MimeAliases {
			setPreferredAssetMimeKind(out, alias, kind)
		}
	}
	return out
}

func mustLoadAssetFileTypes() []assetFileType {
	var items []assetFileType
	if err := json.Unmarshal(assetFileTypesJSON, &items); err != nil {
		panic(fmt.Sprintf("load asset file types: %v", err))
	}
	if len(items) == 0 {
		panic("asset file types cannot be empty")
	}
	seenExts := map[string]bool{}
	for _, item := range items {
		ext := normalizeAssetFileExt(item.Ext)
		mimeType := normalizeAssetMime(item.Mime)
		kind := strings.TrimSpace(item.Kind)
		if ext == "" || mimeType == "" || kind == "" {
			panic(fmt.Sprintf("invalid asset file type: %#v", item))
		}
		if seenExts[ext] {
			panic(fmt.Sprintf("duplicate asset file extension: %s", ext))
		}
		seenExts[ext] = true
	}
	return items
}

func setPreferredAssetMimeKind(target map[string]string, mimeType string, kind string) {
	normalizedMime := normalizeAssetMime(mimeType)
	if normalizedMime == "" || strings.TrimSpace(kind) == "" {
		return
	}
	if _, exists := target[normalizedMime]; exists {
		return
	}
	target[normalizedMime] = strings.TrimSpace(kind)
}

func normalizeAssetFileExt(ext string) string {
	return strings.TrimPrefix(strings.ToLower(strings.TrimSpace(ext)), ".")
}

func normalizeAssetMime(mimeType string) string {
	return strings.TrimSpace(strings.Split(strings.ToLower(strings.TrimSpace(mimeType)), ";")[0])
}
